bindings/rust/bench/historical-perf/bench-past.sh

```diff
@@ -0,0 +1,66 @@
+#!/bin/bash
+
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+# suppress stdout
+exec >/dev/null
+export CARGO_TERM_QUIET=true
+export RUSTFLAGS=-Awarnings
+
+# immediately bail if any command fails
+set -e
+
+# go to s2n-tls/bindings/rust/bench/
+bench_path=`pwd`/`dirname "$0"`/../
+pushd $bench_path
+
+# make Cargo.toml point s2n-tls to the cloned old version
+original_s2n_dep="$(grep 's2n-tls =' Cargo.toml)"
+sed -i "s|s2n-tls = .*|s2n-tls = { path = \"target/s2n-tls/bindings/rust/s2n-tls\" }|" Cargo.toml 
+
+# clone copy of repo to target/s2n-tls
+echo "cloning repo" >&2
+mkdir -p target
+cd target
+rm -rf s2n-tls
+git clone --quiet https://github.com/aws/s2n-tls
+cd s2n-tls/bindings/rust/
+copied_bindings_path=`pwd`
+
+# last tag we want is v1.3.16, get all tags from then
+# `git tag -l | sort -rV` gets list of sorted tags newest to oldest
+
+# get the line number of v1.3.16
+line_num_last_tag=`git tag -l | sort -rV | grep "v1.3.16" --line-number | head -n 1 | cut -d":" -f1`
+
+# loop through all tags in order up to v1.3.16
+for tag in `git tag -l | sort -rV | head -$line_num_last_tag`
+do
+    (
+        # go to s2n-tls/bindings/rust/ inside copied repo
+        cd $copied_bindings_path
+
+        echo "checkout tag $tag" >&2
+        git checkout $tag --quiet
+
+        echo "generating rust bindings" >&2
+        # if generate.sh fails, exit out of block
+        ./generate.sh || exit 1
+
+        cd $bench_path
+        echo "running cargo bench and saving results" >&2
+        cargo bench --features "s2n-only" --no-fail-fast
+        cargo run --release --bin parse_criterion $tag historical-perf/perf.csv
+    ) || echo "failed, trying next tag"
+    echo
+done
+
+# reset Cargo.toml
+cd $bench_path
+sed -i "s|s2n-tls = .*|$original_s2n_dep|" Cargo.toml
```

Comment:
I agree, that was one of my concerns as well at first, and I was originally banking on nothing failing, but that doesn't seem like a good idea.

I played around with adding a `--config` flag to bench, adding the new path to a `[patch]` section, and adding a `paths` override, but none of them work because the original dependency is already a path dependency. (Adding a `paths` override does work for now but because other dependencies change, there's an warning that will become an error in the future.) 

Another way I thought could work but isn't as pretty but works is to `sed` to a new `Cargo.toml` specifically for historical benchmarking, and direct the config to the new `Cargo.toml` with `--manifest-path`. However, that flag requires the name of the new manifest to be `Cargo.toml`, and to not overwrite the existing `Cargo.toml`, you would have to put the modified one in a different directory. Doing this then leads to the problem of all of the paths in the manifest to be messed up, since the paths are relative to the location of the `Cargo.toml`.

Using `sed` in-place on the existing `Cargo.toml` seems to be the best (only?) option so far. Since the things in the for loop in the script are in a `(...) || echo` block, it resists most failures even with with the `set -e` at the beginning. I tried using `exit 1` inside the block, and the script still changed the Cargo.toml back. However, this still didn't work for interrupts with <kbd>Ctrl+C</kbd>. The only way to fix that was to add a `trap` right after the first `sed`, intercepting the exit signals to do the cleanup before exiting.

TL;DR: `sed` still seems like the best/only option, and I added a `trap` to make sure cleanup happens (revert `Cargo.toml`) no matter the exit signal.